src/providers/aws.rs

```diff
@@ -0,0 +1,887 @@
+                                    "iid" => &instance_id,
+                                );
+                            }
+                            _ => {
+                                error!(
+                                    log,
+                                    "spot request failed: {:?}", &sir.status;
+                                    "req_id" => sir.spot_instance_request_id,
+                                );
+                            }
+                        }
+                    }
+                    bail!("wait limit reached");
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Poll AWS until `max_wait` (if not `None`) or the instances are ready to SSH to.
+    pub fn wait_for_instances(&mut self, max_wait: Option<time::Duration>) -> Result<(), Error> {
+        let start = time::Instant::now();
+        let mut desc_req = rusoto_ec2::DescribeInstancesRequest::default();
+        let client = self.client.as_ref().unwrap();
+        let log = self.log.as_ref().unwrap();
+        let private_key_path = self.private_key_path.as_ref().unwrap();
+        let mut all_ready = self.instances.is_empty();
+        desc_req.instance_ids = Some(self.instances.keys().cloned().collect());
+        while !all_ready {
+            all_ready = true;
+
+            for reservation in client
+                .describe_instances(desc_req.clone())
+                .sync()
+                .context("failed to cancel spot instances")?
+                .reservations
+                .unwrap_or_else(Vec::new)
+            {
+                for instance in reservation.instances.unwrap_or_else(Vec::new) {
+                    match instance {
+                        rusoto_ec2::Instance {
+                            state: Some(rusoto_ec2::InstanceState { code: Some(16), .. }),
+                            instance_id: Some(instance_id),
+                            public_dns_name: Some(public_dns),
+                            public_ip_address: Some(public_ip),
+                            ..
+                        } => {
+                            trace!(log, "instance ready";
+                                "instance_id" => instance_id.clone(),
+                                "ip" => &public_ip,
+                            );
+
+                            let (ipinfo, _) = self.instances.get_mut(&instance_id).unwrap();
+                            *ipinfo = Some((public_ip.clone(), public_dns.clone()));
+                        }
+                        _ => {
+                            all_ready = false;
+                        }
+                    }
+                }
+            }
+
+            if let Some(to) = max_wait {
+                if time::Instant::now().duration_since(start) > to {
+                    bail!("timed out");
+                }
+            }
+        }
+
+        use rayon::prelude::*;
+        self.instances
+            .par_iter()
```

Comment:
I think we should document somewhere that regions happen sequentially, but instances within a region are set up in parallel.